decidim-core/db/migrate/20220127113419_add_component_notification_settings_to_users.rb

```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddComponentNotificationSettingsToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :decidim_users, :component_notification_settings, :jsonb, default: {}
```

Comment:
I have some hesitation about adding a jsonb type for this one. Just last week, we were discussing with @quinHD about  #8833, and he mentioned that maybe we're approaching the time to have a notifications table, as this is having more and more entity. I think I'm more inclined with that approach.
The principal concern that I see with having a jsonb on this case is specially when the moment to migrate the `newsletter_notifications` column arrives. We probably could have performance problems by going to all of these jsonb columns to find this. With the current classic approach, if we have this problem we can add an index in this column, but with jsonb it's not that easy. 
I'd also like to hear @ahukkanen input on this one.

Reply:
@andreslucena At the moment when I've started to re-work this task, we didn't discusses to have a notification table. I've choose json  considering the fact that we can have more settings in the future. #8833 is still in progress... I do not know if it's ok for us to wait until further implementations.

Reply:
@andreslucena , this JSONB field has been added to be able to remove the collapase the following fields from User:
- email_on_moderations 
- email_on_notification
- notification_types
- direct_message_types
- email_on_moderations
This also gives the opportunity of adding bigger granularity regarding the notification types, as you could set from the manifest a new notification setting, that would be applicable only for a specific engine (let's say proposals).

Reply:
I think the performance concern @andreslucena is irrelevant with the amount of data that Decidim has in the users table even in the largest instances. I made a small bash script to compare the performance between these two approaches:
https://gist.github.com/ahukkanen/0020604a1c3dbb0deca19ac8c9a001e7

Use it with caution and first read what it does if you want to try it out. Adjust to your needs.

The script is not perfect comparison but I could not get any sensible data out of pg_stat_statements as it basically showed even performance between the two queries. After running it for a while, it doesn't seem to affect the average query performance on how many selects are done. Only the amount of returned data in the table matters regarding the query performance. The tests also have some overhead with the random ordering (to limit the query cache effect) as well as constantly writing the results to /dev/null but I think a similar writing/reading performance also applies to Rails. The only idea is to compare the difference between the two queries when the exact performance of a single query should not matter when comparing two alternative approaches.

I figured there is a maximum of 30% performance difference between the two approaches and the difference gets smaller and smaller as the amount of records increases in the DB. Even with a million records, the base query takes less than a second with both approaches and as said previously, the performance difference is even smaller with higher amount of rows. I don't think there's any single Decidim instance that currently has 1 million users.

So, the point is:
a) If the query takes less than a second, a **maximum** of 30% performance decrease means a third of a second (with a million records), the actual decrease was less than 5% with a million records
b) I think these queries don't happen too often that this difference would make any major difference

So from my perspective, the JSONB column is fine. If the idea is to move all the notification settings to that column, maybe I would rename it as `notification_settings` instead of `component_notification_settings`.

Reply:
And another point is that I would not suggest creating database tables just to hold information about what type of notifications and notification groups there are in Decidim. I think this is rather static information unless we allow admins to disable specific notifications totally.

If there is no need to dynamically manage the notifications or notification groups after Decidim is started, I don't see a reason to put this information to a dedicated table in the DB.

It could be debated whether we should move all user settings to a separate DB table but that's another topic outside of this issue.

Reply:
> So from my perspective, the JSONB column is fine. If the idea is to move all the notification settings to that column, maybe I would rename it as `notification_settings` instead of `component_notification_settings`.

We've wanted the name to reflect that all those settings are related/applicable to a specific component (meetings, proposals etc.) and not to the entire app.

Reply:
@roxanaopr What @alecslupu said above, I understood that you also want to put the current notification settings into this JSONB column when all of them wouldn't be component specific.

Did I understand something wrong?

Reply:
@ahukkanen In the future I think that we can have all the notification settings on the same structure.

Reply:
@ahukkanen , yes, we could rename to `notifications_settings`, and migrate all the other notifications to a registry form like: 

```
      initializer "decidim_core.notification_settings" do
        Decidim.notification_settings(:email_on_moderations) { |ns| ns.settings_area = :administrators }
        Decidim.notification_settings(:email_on_notification) { |ns| ns.settings_area = :general }
      end
```

The refactor would be done at a later stage, after we validate the proper functioning of the close meeting reminders

Reply:
@alecslupu I agree with @ahukkanen, can you rename it to `notification_settings` in this PR please? Thanks!

Reply:
@andreslucena  I've started working on the refactoring part (rename _component_notification_settings_ to _notification_settings_ and include all the other user settings: email_on_notification, email_on_moderations, newsletter_notifications, notifications_from_followed, notifications_from_own_activity, allow_public_contact). The problem is that lots of files are impacted (including spec files, schema migrations, data migrations) and I think that this PR will look very messy at the end. My suggestion is to do this refactoring in a separate PR, to keep thinks more clear.

Reply:
Marking this as solved as the main concerns raised here are handled: performance is not a problem for our current load and the field has been renamed